readthedocs/analytics/models.py

```diff
@@ -33,60 +58,80 @@ class PageView(models.Model):
         verbose_name=_('Version'),
         related_name='page_views',
         on_delete=models.CASCADE,
+        null=True,
     )
     path = models.CharField(max_length=4096)
     view_count = models.PositiveIntegerField(default=0)
     date = models.DateField(default=datetime.date.today, db_index=True)
+    status = models.PositiveIntegerField(
+        default=200,
+        help_text=_("HTTP status code"),
+    )
+
+    objects = PageViewManager()
 
     class Meta:
-        unique_together = ("project", "version", "path", "date")
+        unique_together = ("project", "version", "path", "date", "status")
 
     def __str__(self):
         return f'PageView: [{self.project.slug}:{self.version.slug}] - {self.path} for {self.date}'
 
     @classmethod
-    def top_viewed_pages(cls, project, since=None, limit=10):
+    def top_viewed_pages(cls, project, since=None, limit=10, status=200, per_version=False):
         """
         Returns top pages according to view counts.
 
-        Structure of returned data is compatible to make graphs.
-        Sample returned data::
-        {
-            'pages': ['index', 'config-file/v1', 'intro/import-guide'],
-            'view_counts': [150, 120, 100]
-        }
-        This data shows that `index` is the most viewed page having 150 total views,
-        followed by `config-file/v1` and `intro/import-guide` having 120 and
-        100 total page views respectively.
+        :param per_version: If `True`, group the results per version.
+
+        :returns: A list of named tuples ordered by the number of views.
+         Each tuple contains: path, full_path, and total_views.
         """
         if since is None:
             since = timezone.now().date() - timezone.timedelta(days=30)
 
+        group_by = ["path"]
+        values = ["path", "total_views"]
+        if per_version:
+            group_by.append("version")
+            values.append("version__slug")
+
         queryset = (
-            cls.objects
-            .filter(project=project, date__gte=since)
-            .values_list('path')
-            .annotate(total_views=Sum('view_count'))
-            .values_list('path', 'total_views')
-            .order_by('-total_views')[:limit]
+            cls.objects.filter(project=project, date__gte=since, status=status)
+            .values_list(*group_by)
+            .annotate(total_views=Sum("view_count"))
+            .values_list(*values, named=True)
+            .order_by("-total_views")[:limit]
         )
 
-        pages = []
-        view_counts = []
-
-        for data in queryset.iterator():
-            pages.append(data[0])
-            view_counts.append(data[1])
-
-        final_data = {
-            'pages': pages,
-            'view_counts': view_counts,
-        }
-
-        return final_data
+        PageViewResult = namedtuple('PageViewResult', 'path, full_path, count')
```

Comment:
Probably it makes more sense to store the full path in the model instead of calculating it each time, it will be more precise too.